src/main/java/webserver/routes/TodoRoutes.java

```diff
@@ -1,21 +1,25 @@
 package webserver.routes;
 
-import webserver.JsonTodos;
 import webserver.Router;
 import webserver.controllers.TodoController;
+import webserver.data.TodoService;
 
 public class TodoRoutes {
     TodoController todoController;
     
-    public TodoRoutes() {
-        this.todoController = new TodoController(new JsonTodos("/public/todos.json"));
+    public TodoRoutes(String directory) {
+        this.todoController = new TodoController(new TodoService(directory));
```

Comment:
I think this is the line you said felt the most weird to you, and I can offer some thoughts on why that might be.  But before that, the bottom line is **this can be a fine solution if it's not causing you any problems**.  If your code works, if the structure isn't getting in the way of modifications you plan to make, and if you're not having any trouble testing it -- you're good to go.

You _may_, depending upon future requirements, have difficulty in a couple of areas due to the use of `new` inside another constructor.  It all comes down to the D in SOLID -- dependency inversion.  You don't really _need_ a directory here.  You _need_ a controller for TODO items.  The fact that the controller needs a TodoService and that service needs a directory are all just details.  The more you push those details to the outer-most layer of your code (somewhere in the neighborhood of the `App` class), the easier it is to change.

Specifically, it won't be possible to test `TodoRoutes` without having a real controller that has a real service that stores files in a real directory.  If you can set all that stuff up in a test, **you're still good to go**.  If you want to test this class with a unit test _and_ you don't want to go through all that hassle, then you might wish to express the dependency more directly: have a version of `TodoRoutes::ctor` that takes in the `TodoController`, so you can pass in a mock (for testing) or a different controller (if you need to have different web server behavior) or a controller with a different service (if you need the same behavior as a web server, but want to store TODO items somewhere other than a JSON file in a directory on the filesystem).  _The more directly you express dependencies, the easier they are to update when requirements change_.

There's a middle ground I'm fond of using in these situations - **just have 2 constructors**.  One does all the crap you don't want to think about (the constructor you already have), and the other is more flexible (maybe it's only used in a test for now, but that could change).

```java
public class TodoRoutes {
  public TodoRoutes(String directory) // the constructor you already have; keep using it

  public TodoRoutes(TodoController controller) // use this for testing, or if your requirements change
}
```